arch/arm/src/stm32h7/stm32_spi_slave.c

```diff
@@ -0,0 +1,1796 @@
+#define SPI_SLAVE_INIT_DMA(x)
+#endif
+
+#ifdef CONFIG_PM
+#define SPI_SLAVE_INIT_PM_PREPARE .pm_cb.prepare = spi_pm_prepare,
+#else
+#define SPI_SLAVE_INIT_PM_PREPARE
+#endif
+
+#define SPI_SLAVE_INIT(x)                               \
+{                                                       \
+  .sctrlr.ops    = &g_sctrlr_ops,                       \
+  .sdev          = NULL,                                \
+  .spibase       = STM32_SPI##x##_BASE,                 \
+  .spiclock      = SPI45_KERNEL_CLOCK_FREQ,             \
+  .irq           = STM32_IRQ_SPI##x,                    \
+  SPI_SLAVE_INIT_DMA(x)                                 \
+  .initialized   = false,                               \
+  SPI_SLAVE_INIT_PM_PREPARE                             \
+  .config        = CONFIG_STM32H7_SPI##x##_COMMTYPE,    \
+}
+
+#ifdef CONFIG_STM32H7_SPI1_SLAVE
+
+static uint8_t SPI_SLAVE_OUTQ(1)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static uint8_t SPI_SLAVE_INQ(1)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static struct stm32_spidev_s g_spi1sctrlr = SPI_SLAVE_INIT(1);
+
+#endif
+
+#ifdef CONFIG_STM32H7_SPI2_SLAVE
+
+static uint8_t SPI_SLAVE_OUTQ(2)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static uint8_t SPI_SLAVE_INQ(2)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static struct stm32_spidev_s g_spi2sctrlr = SPI_SLAVE_INIT(2);
+
+#endif
+
+#ifdef CONFIG_STM32H7_SPI3_SLAVE
+
+static uint8_t SPI_SLAVE_OUTQ(3)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static uint8_t SPI_SLAVE_INQ(3)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static struct stm32_spidev_s g_spi3sctrlr = SPI_SLAVE_INIT(3);
+
+#endif
+
+#ifdef CONFIG_STM32H7_SPI4_SLAVE
+
+static uint8_t SPI_SLAVE_OUTQ(4)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static uint8_t SPI_SLAVE_INQ(4)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static struct stm32_spidev_s g_spi4sctrlr = SPI_SLAVE_INIT(4);
+
+#endif
+
+#ifdef CONFIG_STM32H7_SPI5_SLAVE
+
+static uint8_t SPI_SLAVE_OUTQ(5)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static uint8_t SPI_SLAVE_INQ(5)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static struct stm32_spidev_s g_spi5sctrlr = SPI_SLAVE_INIT(5);
+
+#endif
+
+#ifdef CONFIG_STM32H7_SPI6_SLAVE
+
+static uint8_t SPI_SLAVE_OUTQ(6)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
+static uint8_t SPI_SLAVE_INQ(6)[DMA_ALIGN_UP(CONFIG_STM32H7_SPI_SLAVE_QSIZE)]
+__attribute__((aligned(ARMV7M_DCACHE_LINESIZE)));
```

Comment:
Looking at the H7 reference, SPI6 is in the D3 domain. This will require locating this buffer in SRAM3. I feel 'locating' is going to be a more common need. Should we try to formalize it in the compiler.h header?

Reply:
Good catch! I actually never tried to use this on SPI6 so this wasn't noticed. I am putting everything into AXI_SRAM at the moment.

In general, "locating" is indeed a very common need! But I fear that this spreads easily accross startup files, linker scripts and so on.

Maybe for this specific PR, just removing SPI6 support for the slave for now is adequate.

Actually, also for dynamic memory allocations it would be nice to be able to allocate a dma memory pool and malloc memory from there. I do know of gran allocator and "fat_dma_alloc", but those could maybe be more generic than just reserved for fat?

Reply:
I will need to add locating to the master interface as well. So it has to happen sometime.

>Maybe for this specific PR, just removing SPI6 support for the slave for now is adequate.
Sure.

> Actually, also for dynamic memory allocations it would be nice to be able to allocate a dma memory pool and malloc memory from there. I do know of gran allocator and "fat_dma_alloc", but those could maybe be more generic than just reserved for fat?

I imagine having dynamic allocation is really use case dependent. We used the gran allocator  this way in PX4.  I have gone back and forth on this. But in reality you have to pre-allocate the memory  in a static chunk, and size it for the max utilization.  So why not just statically allocated the memory in the driver, per channel's worst case need? At the end of the day it is simpler and the amount of memory allocated is == (or less if you over allocate for margin in the gran allocator)